pandas/_testing/asserters.py

```diff
@@ -447,6 +450,15 @@ def repr_class(x):
         return type(x).__name__
 
     if exact == "equiv":
+        ldtype = left.dtype
+        rdtype = right.dtype
+
+        if is_numeric_dtype(ldtype) and is_numeric_dtype(rdtype):
```

Comment:
wouldn't this obviate the logic for below? are we actually hitting the below logic after this change?

Reply:
This is ok AFAICS, but should be located in _check_types.

I've made a PR #41960 dto do some cleanup, before moving this.

Reply:
I've refactored this in #41980, so this file isn't touched in this PR any more.